### DISCUSSION TITLE
GDS - Reestablish UART connection

### DISCUSSION THREAD

**Discussion by arturodrt** [nasa/fprime#1882] [Q&A] [answered]
Hello, 

Does anyone know if it's possible to reset a UART connection once it's been lost? When I do a hard reset on my GPS sensor it disconnects from the UART and to reconnect it I have to restart the GDS server. I am wondering if there is a way to implement a function or a command to establish connection again without closing the GDS server.

![uart](https://user-images.githubusercontent.com/63877617/217301148-3bb94446-f096-421a-94ce-642cecee96e3.png)

**Comment by LeStarch** [comment, marked as answer]
This is likely happening in the flight software itself and the display is only done on the GDS.  You can look into the UART component for your system to see if it retries connection on error.

**Reply by arturodrt** [reply]
![image](https://user-images.githubusercontent.com/63877617/217872485-9a6d3ee8-3eab-44bc-bbf2-3b3da8c0390b.png)

I am working with this configuration for UART communication as per RPI example. 

After GPS restart, I am trying to reestablish connection with the following LinuxSerialDriverComponentImpl methods:
open()
startReadThread()
quitReadThread()

But I have not been able to send messages to the GPS device. Even when I execute the open() method with return **status = true**. When I try to write a data buffer I continue falling into this path ("Error Writing UART device").

![image](https://user-images.githubusercontent.com/63877617/217876042-6297762d-3a64-497a-a899-23caf081bf4e.png)

**Reply by LeStarch** [reply]
How are you triggering a re-connection?  The Serial driver does not attempt to reconnect after error, so some part of your code would need to handle that.

**Reply by arturodrt** [reply]
Well I am currently trying to figure out how to trigger the re-connection. My plan is to use the "close(this->m_fd)" function to close the UART to then open it again with a new file descriptor because after the UART disconnection if I want to use the read and write functions they always return -1. 

(by UART disconnection I mean that the gps device is powered off for a short period of time and then it is available again on the same serial port.)

**Reply by arturodrt** [reply]
It worked :D! Thank you for your answers!

**Reply by LeStarch** [reply]
What worked? `close` then `open`?  Just want that captured here if possible.

**Reply by arturodrt** [reply]
Yes, when I can no longer write or read, I use the close function and then exit the thread to avoid segmentation fault and core dump errors. Once this is done a UART connection can be re-opened, what I did for now was create a custom command that I can execute whenever I want to re-establish the connection again to the serial port.
